### PR TITLE
geomfn: ensure NaN never intersects/covers/within

### LINKED ISSUE #79992: sql: v21.2, v22.1 incorrect results for st_intersects filter
roachtest.tlp [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4897769&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4897769&tab=artifacts#/tlp) on release-22.1.0 @ [ab73f1865cd4ed6812cb413ab5d45aed6a0644dc](https://github.com/cockroachdb/cockroach/commits/ab73f1865cd4ed6812cb413ab5d45aed6a0644dc):

```
The test failed on branch=release-22.1.0, cloud=gce:
test artifacts and logs in: /artifacts/tlp/run_1
	tlp.go:172,tlp.go:76,test_runner.go:875: expected unpartitioned and partitioned results to be equal
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runTLPQuery.func2
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/tlp.go:245
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runWithTimeout.func1
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/tlp.go:270
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) expected unpartitioned and partitioned results to be equal
		  |   []string{
		  |   	... // 19 identical elements
		  |   	"2389279048,2946281655,3174868576,2028-11-12 19:49:46.000359 +000"...,
		  |   	"2389279048,2946281655,3174868576,2028-11-12 19:49:46.000359 +000"...,
		  | + 	"2389279048,2946281655,3174868576,2028-11-12 19:49:46.000359 +0000 +0000,1695361665,{p,X,?.1eV,,105,0000-01-01 09:06:52.217956 +0000 UTC,0000-01-01 14:33:38.295388 +0002 +0002,NaN,2024-09-23 22:57:23.000022 +0000 +0000,NULL, \x04,ec08e142-77b0-49b0-9bdd-83a394"...,
		  |   	"2389279048,2946281655,3174868576,2028-11-12 19:49:46.000359 +000"...,
		  |   	"2389279048,2946281655,3174868576,2028-11-12 19:49:46.000359 +000"...,
		  |   	... // 40 identical elements
		  |   	"2737886472,1454533650,NULL,0001-01-01 00:00:00 +0000 +0000,0,{p,"...,
		  |   	"2737886472,1454533650,NULL,0001-01-01 00:00:00 +0000 +0000,0,{p,"...,
		  | + 	"2737886472,1454533650,NULL,0001-01-01 00:00:00 +0000 +0000,0,{p,X,?.1eV,,105,0000-01-01 09:06:52.217956 +0000 UTC,0000-01-01 14:33:38.295388 +0002 +0002,NaN,2024-09-23 22:57:23.000022 +0000 +0000,NULL, \x04,ec08e142-77b0-49b0-9bdd-83a394cc2f4c,☃,T),:z\x02fO\x0fA,"...,
		  |   	"2737886472,1454533650,NULL,0001-01-01 00:00:00 +0000 +0000,0,{p,"...,
		  |   	"2737886472,1454533650,NULL,0001-01-01 00:00:00 +0000 +0000,0,{p,"...,
		  |   	... // 23 identical elements
		  |   }
		  | sql: SELECT * FROM defaultdb.public.table3 AS tab_11451 JOIN defaultdb.public.table2 AS tab_11452 ON true
		  | (SELECT * FROM defaultdb.public.table3 AS tab_11451 JOIN defaultdb.public.table2 AS tab_11452 ON st_disjoint('01030000C0010000000B000000DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1A09A237582CECA41F867897365D0D14109CF7245217E00C21FCDDF330C6CFBC1A09C6362F99AFD41357938A7E23F01C29BAD8985F5F3FDC1E6822AC5CFD4F741F47C7067330DF44188CB2A2EB416D2C1C7921866C903F5C1306BDF05C89AE0416AC95CF8DC340242C0C6AABD892AB5C1309FD65830F3C241908A1A9F3134E0C1B2564E4DA9C7F841A8442BAA7D95D04115962ACE402A00C26E77E75507FBF0410CC3B58EFCBDFB4118634FDC6BAD0042134F1D836D48FFC1D44EDF69D768E0C1000156A00179CEC19C3804536204FF410B03A1C6FB7DF5C19ACFC550DC4DFDC18AB3BABC47FDE1C1704D48C6A1710242B6ACDBC1FEF3EEC1B29BF7DB4610F541C9A4654CCDD1F2C1E024CF7490C5E34180D06C8E1F7DB641A4B9C124F8C0E541DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1'::GEOMETRY, st_makepoint(tab_11452.col2_3::FLOAT8, 0.7853641330683674::FLOAT8, -0.8997400826112889::FLOAT8)::GEOMETRY::GEOMETRY)::BOOL) UNION ALL (SELECT * FROM defaultdb.public.table3 AS tab_11451 JOIN defaultdb.public.table2 AS tab_11452 ON NOT (st_disjoint('01030000C0010000000B000000DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1A09A237582CECA41F867897365D0D14109CF7245217E00C21FCDDF330C6CFBC1A09C6362F99AFD41357938A7E23F01C29BAD8985F5F3FDC1E6822AC5CFD4F741F47C7067330DF44188CB2A2EB416D2C1C7921866C903F5C1306BDF05C89AE0416AC95CF8DC340242C0C6AABD892AB5C1309FD65830F3C241908A1A9F3134E0C1B2564E4DA9C7F841A8442BAA7D95D04115962ACE402A00C26E77E75507FBF0410CC3B58EFCBDFB4118634FDC6BAD0042134F1D836D48FFC1D44EDF69D768E0C1000156A00179CEC19C3804536204FF410B03A1C6FB7DF5C19ACFC550DC4DFDC18AB3BABC47FDE1C1704D48C6A1710242B6ACDBC1FEF3EEC1B29BF7DB4610F541C9A4654CCDD1F2C1E024CF7490C5E34180D06C8E1F7DB641A4B9C124F8C0E541DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1'::GEOMETRY, st_makepoint(tab_11452.col2_3::FLOAT8, 0.7853641330683674::FLOAT8, -0.8997400826112889::FLOAT8)::GEOMETRY::GEOMETRY)::BOOL)) UNION ALL (SELECT * FROM defaultdb.public.table3 AS tab_11451 JOIN defaultdb.public.table2 AS tab_11452 ON (st_disjoint('01030000C0010000000B000000DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1A09A237582CECA41F867897365D0D14109CF7245217E00C21FCDDF330C6CFBC1A09C6362F99AFD41357938A7E23F01C29BAD8985F5F3FDC1E6822AC5CFD4F741F47C7067330DF44188CB2A2EB416D2C1C7921866C903F5C1306BDF05C89AE0416AC95CF8DC340242C0C6AABD892AB5C1309FD65830F3C241908A1A9F3134E0C1B2564E4DA9C7F841A8442BAA7D95D04115962ACE402A00C26E77E75507FBF0410CC3B58EFCBDFB4118634FDC6BAD0042134F1D836D48FFC1D44EDF69D768E0C1000156A00179CEC19C3804536204FF410B03A1C6FB7DF5C19ACFC550DC4DFDC18AB3BABC47FDE1C1704D48C6A1710242B6ACDBC1FEF3EEC1B29BF7DB4610F541C9A4654CCDD1F2C1E024CF7490C5E34180D06C8E1F7DB641A4B9C124F8C0E541DE548D8493D5FAC198BAF6E9D615D7C16F8C4AC1CD1C00C27244AB6FFC99EDC1'::GEOMETRY, st_makepoint(tab_11452.col2_3::FLOAT8, 0.7853641330683674::FLOAT8, -0.8997400826112889::FLOAT8)::GEOMETRY::GEOMETRY)::BOOL) IS NULL)
		  | with args: []
		Error types: (1) *withstack.withStack (2) *errutil.leafError
```
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #79947 roachtest: tlp failed [C-test-failure O-roachtest O-robot T-sql-queries branch-release-22.1 release-blocker]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tlp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by otan**
Resolves #79992

Release note (bug fix): Fixed a bug where NaN coordinates when using
ST_Intersects/ST_Within/ST_Covers would return true instead of false for
point in polygon operations.

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
Does this fix the test case in #79992? Or is it possible that tlp will continue to find this as an issue?

If it does fix it, can you backport the fix to 22.1 and 21.2? Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80136#-:-N00njy8EnXyr0W90qPu:bpsx28i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

File: pkg/geo/geomfn/topology_operations.go [hunk 1]
```diff
@@ -208,17 +208,36 @@ func MinimumRotatedRectangle(g geo.Geometry) (geo.Geometry, error) {
...
+
+// BoundingBoxHasNaNCoordinates checks if the bounding box of a Geometry
+// has a NaN coordinate.
+func BoundingBoxHasNaNCoordinates(g geo.Geometry) bool {
+	boundingBox := g.BoundingBoxRef()
+	if boundingBox == nil {
+		return false
+	}
+	// Don't use `:= range []float64{...}` to avoid memory allocation.
+	isNaN := func(ord float64) bool {
```
Comment (rafiss) [inline pkg/geo/geomfn/topology_operations.go:236]: very advanced helper function :D

**Review by rafiss — APPROVED** [review]

**Comment by otan** [comment]
> If it does fix it, can you backport the fix to 22.1 and 21.2? Thanks!

it does, but i need to backport two PRs so that i don't get a merge conflict so i'll do it by hand when this merges

bors r=rafiss

**Comment by otan** [comment]
oh wait that commit that may merge conflict is already merged, blathers should do it

**Pull request merged**